File: Sorting Report/QuickSort/QuickSort.py
from time import *

# Read Input
with open('input.txt', 'r') as file:
    lines = file.readlines()
    arrs = []
    for i in range(0, len(lines), 2):
        n = int(lines[i].strip())
        arrs.append(list(map(int, lines[i + 1].strip().split())))

# Quick Sort
def QuickSort(arr : list, low : int, high : int):
    pivot = arr[(low + high) // 2]
    i, j = low, high

    while i < j:
        while arr[i] < pivot:
            i += 1
        while arr[j] > pivot:
            j -= 1

        if i <= j:
            arr[i], arr[j] = arr[j], arr[i]  # Swap
            i += 1
            j -= 1

    if i < high:
        QuickSort(arr, i, high)
    if j > low:
        QuickSort(arr, low, j)

# Sort and Check time
for arr in arrs:
    start_time = time()
    QuickSort(arr, 0, len(arr) - 1)
    finish_time = time()

    print(f"Array (Sorted in {finish_time - start_time:.6f} seconds)")

File: Exercise 1/QuickSort/QuickSort.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <sstream>
#include <chrono>

using namespace std;
using namespace chrono;

// Quick Sort
void QuickSort(vector<int>& arr, int low, int high) {
    if (low >= high) return;

    int pivot = arr[(low + high) / 2];
    int i = low, j = high;

    while (i <= j) {
        while (arr[i] < pivot) i++;
        while (arr[j] > pivot) j--;

        if (i <= j)
            swap(arr[i++], arr[j--]);
    }

    if (low < j) QuickSort(arr, low, j);
    if (i < high) QuickSort(arr, i, high);
}

int main() {
    ifstream inputFile("input.txt");
    vector<vector<int>> arrs;
    string line;

    // Read Input
    while (getline(inputFile, line)) {
        int n = stoi(line), num;
        vector<int> arr;

        if (getline(inputFile, line)) {
            stringstream ss(line);
            while (ss >> num)
                arr.push_back(num);
        }

        arrs.push_back(arr);
    }

    inputFile.close();

    // Sort and Check Time
    for (int i = 0; i < arrs.size(); i++) {
        auto start = high_resolution_clock::now();
        QuickSort(arrs[i], 0, arrs[i].size() - 1);
        auto end = high_resolution_clock::now();

        duration<double> exec_time = end - start;

        cout << "Array " << i + 1 << " (Sorted in " << exec_time.count() << " seconds)" << endl;
    }

    return 0;
}
